### DISCUSSION TITLE
MSBuild on Jenkins WIN node fails "randomly" with MSB3021 error.

### DISCUSSION THREAD

**Discussion by gowindu123** [dotnet/msbuild#10058] [Q&A]
Hello Team,

We are facing the issue of build failure due to MSB3021

We have checked for file locks (like Anti Virus and others) but could notice no process is blocking the files.

After referring the discussions: we have tried setting 

**set MSBUILDALWAYSRETRY=1**
 
_The command that we are invoking for the build is:_ 
"C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Current\Bin\msbuild" ***********.sln /t:Rebuild /p:Configuration=Release /p:Platform="Any CPU" /nodeReuse:false /verbosity:minimal /p:PrecompileBeforePublish=false /p:OutDir=build  /p:DeployTarget=Package /p:DebugType=None /p:DeployOnBuild=true /p:SkipInvalidConfigurations=true /p:PackageAsSingleFile=false /p:AspNetCompileMerge=true /p:RunCodeAnalysis=false /p:WarnAsError=false /nowarn:MSB3247,MSB3277,MSB3021 
**MSBuild version 17.7.2+d6990bcfa for .NET Framework**

The error we get is :
error MSB3021: Unable to copy file ***********.dll" to ******.dll". Access to the path '********.dll' is denied.
[  and this is happening randomly for different dlls in different runs .

We are using VS2022 build toolkit for the solution build and the application is framework application <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>

Please help on this as it is heavily affecting out CI builds

**Comment by KalleOlaviNiemitalo** [comment]
> We have checked for file locks (like Anti Virus and others) but could notice no process is blocking the files.

Did you check during the build, or after the build?

**Reply by gowindu123** [reply]
We have checked during the build and found no files have been impacted

**Reply by gowindu123** [reply]
the process is ending abruptly with the MSB3021 error though no file is being locked

**Reply by GangWang01** [reply]
@gowindu123 the single error message could not help us understand if the copy task with retries worked or not with setting `set MSBUILDALWAYSRETRY=1`. Can you share more detailed build output?
And it sounds like you are able to check files' status when the issue is happening. Can you refer to https://github.com/dotnet/msbuild/issues/9169#issuecomment-1717360053 to provide process monitor log while the issue is happening? It can help us a lot understand why access to the file is denied. It's great if you can provide the binary log for the investigation. But before sharing binary log be aware that binary log may capture sensitive information.
